feat: parse binary response into logs.

BenjaminMummery opened the pull request:
[ORQSDK-652]

# The problem

The client doesn't deserialise get workflow logs responses before returning them.

# This PR's solution

Decodes the responses as a list of strings on the client side. CE_runtime handles reformatting these as a dict of lists of log strings indexed by a sentinel task invocation ID, as we cannot yet link CE logs to a. task invocation.

# Checklist

- [X] Tests have been added for new features/changed behaviour (if no new features have been added, check the box).
- [X] The [changelog file](CHANGELOG.md) has been updated with a user-readable description of the changes (if the change isn't visible to the user in any way, check the box).
- [X] The PR's title is prefixed with `<feat/fix/chore/internal/docs>[!]:`


[ORQSDK-652]: https://zapatacomputing.atlassian.net/browse/ORQSDK-652?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

BenjaminMummery committed:
feat: parse binary response into logs.

[ORQSDK-652]

BenjaminMummery committed:
Merge branch 'main' into ORQSDK-652-client-deserialisation-of-get-workflow-logs-response

BenjaminMummery committed:
tests: add test coverage for exception raising.

[ORQSDK-652]

BenjaminMummery committed:
docs: update changelog.

[ORQSDK-652]

BenjaminMummery committed:
Merge branch 'main' into ORQSDK-652-client-deserialisation-of-get-workflow-logs-response

BenjaminMummery committed:
Merge branch 'main' into ORQSDK-652-client-deserialisation-of-get-workflow-logs-response

BenjaminMummery committed:
Merge branch 'main' into ORQSDK-652-client-deserialisation-of-get-workflow-logs-response

alexjuda reviewed (changes requested):
Left some suggestions and highlighted some things I worry about. I spent a lot of time in the past weeks bugfixing logs in the local runtime, I'm happy to compare notes if needed!

BenjaminMummery committed:
refactor: one-line logs_dict parsing

Co-authored-by: Alexander Juda <6004040+alexjuda@users.noreply.github.com>

BenjaminMummery committed:
feat: remove WorkflowRunLog - using WFLog instead.

[ORQSDK-652]

BenjaminMummery commented on src/orquestra/sdk/_base/_driver/_client.py:
```diff
@@ -579,15 +590,56 @@ def get_workflow_run_logs(self, wf_run_id: _models.WorkflowRunID) -> bytes:
+        for section in decoded.split("\n"):
+            if len(section) < 1:
+                continue
+            for log in json.loads(section):
+                try:
+                    # Orquestra logs are jsonable - where we can we parse these and
+                    # extract the useful information
+                    parsed_log_message = json.loads(log[1]["log"])
```
Good idea! [0968c89](https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/79/commits/0968c890017a6715cc0436fe32cc9d10c8b6be1a) switches to using `WFLog`.

BenjaminMummery replied:
Okay, I think rather than implementing another pydantic model we should maybe leave this as-is for now and push ahead with getting a new logs spec agreed with the platform. [ORQSDK-777](https://zapatacomputing.atlassian.net/browse/ORQSDK-777?atlOrigin=eyJpIjoiYjNjZDg2ODk4YTA4NDI3OWEyYzMzODRkZWYyZWM0ZTYiLCJwIjoiaiJ9) is set up to track further efforts.

BenjaminMummery commented on src/orquestra/sdk/_base/_driver/_ce_runtime.py:
```diff
@@ -323,8 +324,43 @@ def list_workflow_runs(
+        except _exceptions.WorkflowRunLogsNotReadable as e:
+            raise exceptions.InvalidWorkflowRunLogsError(f"{e}") from e
+
+        # TODO: index by taskinvocationID rather than workflowrunID [ORQSDK-777]
+        logs_dict = {}
```
Using None thanks to the spec of WFLog [0968c89](https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/79/commits/0968c890017a6715cc0436fe32cc9d10c8b6be1a)

BenjaminMummery commented on src/orquestra/sdk/schema/workflow_run.py:
```diff
@@ -64,3 +64,15 @@ class WorkflowRun(WorkflowRunMinimal):
+    message: str
+    wf_id: WorkflowRunId
+    task_id: t.Optional[TaskInvocationId]
+    _content: dict
```
As of [0968c89](https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/79/commits/0968c890017a6715cc0436fe32cc9d10c8b6be1a) it doesn't exist any more XD

BenjaminMummery commented on src/orquestra/sdk/_base/_driver/_client.py:
```diff
@@ -579,15 +590,56 @@ def get_workflow_run_logs(self, wf_run_id: _models.WorkflowRunID) -> bytes:
+                        timestamp=log[0],
+                        message=parsed_log_message["message"],
+                        wf_id=parsed_log_message["wf_run_id"],
+                        task_id=parsed_log_message["task_run_id"],
```
Good idea, [0968c89](https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/79/commits/0968c890017a6715cc0436fe32cc9d10c8b6be1a) changes it to use WFLog and removes WorkflowRunLog altogether.

BenjaminMummery committed:
test: update tests for change from WorkflowRunLog to WFLog.

[ORQSDK-652]

BenjaminMummery commented on src/orquestra/sdk/_base/_driver/_ce_runtime.py:
```diff
@@ -323,8 +324,43 @@ def list_workflow_runs(
+                f"Workflow run with id `{wf_run_id}` not found"
+            ) from e
+        except (_exceptions.InvalidTokenError, _exceptions.ForbiddenError) as e:
+            raise exceptions.UnauthorizedError(f"{e}") from e
+        except _exceptions.WorkflowRunLogsNotReadable as e:
+            raise exceptions.InvalidWorkflowRunLogsError(f"{e}") from e
```
Yep, I agree that that's a poor way of doing things. While I'm changing it, shall I also change the following? 

https://github.com/zapatacomputing/orquestra-workflow-sdk/blob/bdeb0be498e1a95b1cf88604eb615211db9517e5/src/orquestra/sdk/_base/_driver/_ce_runtime.py#L115-L116

https://github.com/zapatacomputing/orquestra-workflow-sdk/blob/bdeb0be498e1a95b1cf88604eb615211db9517e5/src/orquestra/sdk/_base/_driver/_ce_runtime.py#L121-L122

https://github.com/zapatacomputing/orquestra-workflow-sdk/blob/bdeb0be498e1a95b1cf88604eb615211db9517e5/src/orquestra/sdk/_base/_driver/_ce_runtime.py#L158-L159

https://github.com/zapatacomputing/orquestra-workflow-sdk/blob/bdeb0be498e1a95b1cf88604eb615211db9517e5/src/orquestra/sdk/_base/_driver/_ce_runtime.py#L201-L202

https://github.com/zapatacomputing/orquestra-workflow-sdk/blob/bdeb0be498e1a95b1cf88604eb615211db9517e5/src/orquestra/sdk/_base/_driver/_ce_runtime.py#L216-L217

https://github.com/zapatacomputing/orquestra-workflow-sdk/blob/bdeb0be498e1a95b1cf88604eb615211db9517e5/src/orquestra/sdk/_base/_driver/_ce_runtime.py#L250-L251

https://github.com/zapatacomputing/orquestra-workflow-sdk/blob/bdeb0be498e1a95b1cf88604eb615211db9517e5/src/orquestra/sdk/_base/_driver/_ce_runtime.py#L296-L297

https://github.com/zapatacomputing/orquestra-workflow-sdk/blob/bdeb0be498e1a95b1cf88604eb615211db9517e5/src/orquestra/sdk/_base/_driver/_ce_runtime.py#L323-L324

BenjaminMummery replied:
Alrighty then :D

BenjaminMummery commented on src/orquestra/sdk/schema/workflow_run.py:
```diff
@@ -64,3 +64,15 @@ class WorkflowRun(WorkflowRunMinimal):
+
+    timestamp: str
+    message: str
+    wf_id: WorkflowRunId
```
Addressed in [0968c89](https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/79/commits/0968c890017a6715cc0436fe32cc9d10c8b6be1a)

BenjaminMummery commented on src/orquestra/sdk/schema/workflow_run.py:
```diff
@@ -64,3 +64,15 @@ class WorkflowRun(WorkflowRunMinimal):
+    A single log from the workflow.
+    """
+
+    timestamp: str
```
All timestamps are now converted to timezone-aware datetime objects. [0968c89](https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/79/commits/0968c890017a6715cc0436fe32cc9d10c8b6be1a)

BenjaminMummery committed:
fix: handle construction of log dict where the log has no run id.

[ORQSDK-652]

BenjaminMummery committed:
test: update ce_runtime tests for change from WorkflowRunLog to WFLog.

[ORQSDK-652]

alexjuda commented on src/orquestra/sdk/_base/_driver/_client.py:
```diff
@@ -579,15 +586,61 @@ def get_workflow_run_logs(self, wf_run_id: _models.WorkflowRunID) -> bytes:
+                    # Orquestra logs are jsonable - where we can we parse these and
+                    # extract the useful information
+                    parsed_log_message = json.loads(log[1]["log"])
+                    interpreted_log = WFLog(
+                        timestamp=datetime.fromtimestamp(log[0], timezone.utc),
+                        level=parsed_log_message["level"],
+                        filename=parsed_log_message["filename"],
+                        message=parsed_log_message["message"],
+                        wf_run_id=parsed_log_message["wf_run_id"],
+                        task_inv_id=None,
+                        task_run_id=parsed_log_message["task_run_id"],
+                    )
```
Let's use one of your examples.
```json
[
  1678804402.015901,
  {
    "tag":"workflow.logs.ray.wf_def-FfnTz-r000",
    "ray_filename":"/tmp/ray/session_latest/logs/worker-a601931abcd04110b31760f245e4c61d1cbd00f5943fabd5eeca1d9e-01000000-235.err",
    "log":"{\"timestamp\": \"2023-03-14T14:32:57.935535+00:00\", \"level\": \"ERROR\", \"filename\": \"_dag.py:195\", \"message\": \"Traceback (most recent call last):\\n  File \\\"/home/orquestra/venv/lib/python3.9/site-packages/orquestra/sdk/_ray/_dag.py\\\", line 192, in _ray_remote\\n    return wrapped(*inner_args, **inner_kwargs)\\n  File \\\"/home/orquestra/venv/lib/python3.9/site-packages/orquestra/sdk/_ray/_dag.py\\\", line 147, in __call__\\n    return self._fn(*unpacked_args, **unpacked_kwargs)\\n  File \\\"/Users/benjaminmummery/Documents/Projects/orquestra-workflow-sdk/../scratch/scratch.py\\\", line 12, in do_thing\\nException: Blah\\n\", \"wf_run_id\": \"wf.wf_def.3c5f938\", \"task_run_id\": \"invocation-0-task-do-thing\"}"
  }
]
```

The `"log"` above should 1:1 correspond to `WFLog`* (that's the purpose of `WFLog` class), so we can replace all the JSON accesses with just:
```suggestion
                    wf_log_json = log[1]["log"]
                    interpreted_log = WFLog.parse_json(wf_log_json)
```
Pydantic models excel at parsing JSON strings! Bonus point: this fixes the bug with `task_inv_id=None,` (most log lines should have it + we're using this field for filtering later on)

(*) disclaimer: the above is true for log lines produced by [our structured logger](https://github.com/zapatacomputing/orquestra-workflow-sdk/blob/f5defc69058644d2044a0de04f23447b3e7cda39/src/orquestra/sdk/_base/_log_adapter.py#L20)

BenjaminMummery replied:
Ah, that's much nicer, thanks!

BenjaminMummery replied:
[3ad8f0e](https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/79/commits/3ad8f0e1781a4459dc2182c800ca3095b9d128ad)

BenjaminMummery commented on src/orquestra/sdk/_base/_driver/_client.py:
```diff
@@ -579,15 +586,61 @@ def get_workflow_run_logs(self, wf_run_id: _models.WorkflowRunID) -> bytes:
+                        task_run_id=parsed_log_message["task_run_id"],
+                    )
+
+                except (json.JSONDecodeError, KeyError):
+                    # If the log isn't jsonable (i.e. it's from Ray) extract what info
+                    # we can.
+                    interpreted_log = WFLog(
+                        timestamp=datetime.fromtimestamp(log[0], timezone.utc),
+                        level="",
+                        filename=log[1]["ray_filename"],
+                        message=log[1]["log"],
+                        wf_run_id=log[1]["tag"],
+                        task_inv_id=None,
+                        task_run_id=None,
+                    )
```
> please don't use empty strings (level="")

I take it you didn't intend to suggest it as a sentinel value then? https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/79#discussion_r1141950401 WFLog won't allow a Nonetype there so I'll switch it to a "FIXME"

> our interpretation of WFLog.filename is the name of the file where the logger was used. ray_filename is probably the name of the log file produced by Ray, which is a different string.

Yep good point, that needs to go!

log[1]["tag"] depends on the fluentbit configuration. Are you sure this is always going to be WorkflowRunId?

No, not even slightly, but I can't think of a way to determine from the information in the logs what it actually is!

alexjuda replied:
ATM this method produces `WFLog`s. Seems like the only field we're using it for (as of today) is the `message`: https://github.com/zapatacomputing/orquestra-workflow-sdk/blob/c8387d76750ab892454b7d953a07ce7bfeb49acb/src/orquestra/sdk/_base/_driver/_ce_runtime.py#L400

Instead of producing an object with sentinel values, only to discard it soon after, WDYT about just returning a list of strings from this method? See also a patch suggestion in the other comment

BenjaminMummery replied:
[b0e2a54](https://github.com/zapatacomputing/orquestra-workflow-sdk/pull/79/commits/b0e2a54e8cc345eddb9a5a6ebd4ba307a8ebcc99)

BenjaminMummery committed:
Update src/orquestra/sdk/_base/_driver/_client.py

Co-authored-by: Alexander Juda <6004040+alexjuda@users.noreply.github.com>

BenjaminMummery committed:
refactor: remove serialisaion of exceptions.

[ORQSDK-652]

BenjaminMummery committed:
Refactor: use parse_json to interpret jsonable logs

Co-authored-by: Alexander Juda <6004040+alexjuda@users.noreply.github.com>

BenjaminMummery committed:
refactor: no empty strings.

[ORQSDK-652]

BenjaminMummery committed:
feat: ignore existing task inv ids.

[ORQSDK-652]
We don't have task inv ids for ray logs, so if we used them we'd end up splitting ray and sdk logs. This is a temporary measure until we can get task inv ids for those logs.

BenjaminMummery committed:
refactor: use pydantic to parse jsonable logs.

[ORQSDK-652]

BenjaminMummery commented on src/orquestra/sdk/_base/_driver/_client.py:
```diff
@@ -579,15 +586,61 @@ def get_workflow_run_logs(self, wf_run_id: _models.WorkflowRunID) -> bytes:
+
+        untarred = TarFile(fileobj=io.BytesIO(unzipped)).extractfile("step-logs")
+        assert untarred is not None
+        decoded = r"""{}""".format(untarred.read().decode("utf-8"))
```
Nothing at all, that's a holdover from when I was trying to figure out why everything was broken for a while, thanks for spotting it!

BenjaminMummery committed:
fix: remove interpreting ray log filename as WFLog filename.

[ORQSDK-652]

BenjaminMummery committed:
Merge branch 'main' into ORQSDK-652-client-deserialisation-of-get-workflow-logs-response

alexjuda commented on src/orquestra/sdk/_base/_driver/_client.py:
```diff
@@ -579,15 +586,51 @@ def get_workflow_run_logs(self, wf_run_id: _models.WorkflowRunID) -> bytes:
+                try:
+                    # Orquestra logs are jsonable - where we can we parse these and
+                    # extract the useful information
+                    interpreted_log = WFLog.parse_raw(log[1]["log"])
+                except pydantic.ValidationError:
+                    # If the log isn't jsonable (i.e. it's from Ray) extract what info
+                    # we can.
+                    interpreted_log = WFLog(
+                        timestamp=datetime.fromtimestamp(log[0], timezone.utc),
+                        level="UNKNOWN LEVEL",
+                        filename="UNKNOWN FILENAME",
+                        message=log[1]["log"],
+                        wf_run_id=log[1]["tag"],
+                        task_inv_id=None,
+                        task_run_id=None,
+                    )
```
WDYT about this?
```suggestion
                    interpreted_log = WFLog.parse_raw(log[1]["log"])
                    return interpreted_log.message
                except pydantic.ValidationError:
                    # If the log isn't jsonable (i.e. it comes from Ray) we just return 
                    # plain log content.
                    return log[1]["log"]
```

BenjaminMummery committed:
feat: return strings rather than WFLogs

Co-authored-by: Alexander Juda <6004040+alexjuda@users.noreply.github.com>

BenjaminMummery committed:
refactor: change from List[WFLog] to List[str] for ce logs

[ORQSDK-652]

BenjaminMummery committed:
Merge branch 'main' into ORQSDK-652-client-deserialisation-of-get-workflow-logs-response

BenjaminMummery committed:
test: update logs test for change frm WFLog to str

[ORQSDK-652]

alexjuda commented on tests/sdk/v2/driver/test_client.py:
```diff
@@ -1647,12 +1728,29 @@ def test_not_found(
             ):
                 endpoint_mocker(
-                    json=resp_mocks.make_get_wf_run_logs_response(),
+                    # json=resp_mocks.make_get_wf_run_logs_response(),
```
```suggestion
```

alexjuda commented on tests/sdk/v2/driver/test_client.py:
```diff
@@ -1609,7 +1689,8 @@ def test_params_encoding(
                 """
                 endpoint_mocker(
-                    json=resp_mocks.make_get_wf_run_logs_response(),
+                    # json=resp_mocks.make_get_wf_run_logs_response(),
```
```suggestion
```

alexjuda reviewed (approved):
LGTM! Thanks for patience :)
